### PR TITLE
Add size checks for sparse tensor constructor

### LINKED ISSUE #4093: Crash with invalid indices to a sparse matrix
Hi,

```python
1/torch.sparse.FloatTensor((torch.randperm(1000)%10).unsqueeze(0), torch.ones(1000), torch.Size((4,))).to_dense()
```
seems to crash 0.3.0 post 4 on python 3.6.

Best regards

Thomas

Stack trace:
```
#1  0x00007ffff6ce819a in __GI_abort () at abort.c:89
#2  0x00007ffff6d25310 in __libc_message (do_abort=do_abort@entry=2, fmt=fmt@entry=0x7ffff6e1fd30 "*** Error in `%s': %s: 0x%s ***\n") at ../sysdeps/posix/libc_fatal.c:175
#3  0x00007ffff6d2b96e in malloc_printerr (action=3, str=0x7ffff6e1c838 "corrupted double-linked list", ptr=<optimized out>, ar_ptr=<optimized out>) at malloc.c:5079
#4  0x00007ffff6d2bd18 in malloc_consolidate (av=av@entry=0x7ffff7050b00 <main_arena>) at malloc.c:4192
#5  0x00007ffff6d2d7ca in _int_malloc (av=av@entry=0x7ffff7050b00 <main_arena>, bytes=bytes@entry=1528) at malloc.c:3452
#6  0x00007ffff6d2f874 in __GI___libc_malloc (bytes=1528) at malloc.c:2892
#7  0x00007fffb2a722b9 in ?? () from /usr/local/lib/python3.6/dist-packages/torch/lib/libgomp.so.1
#8  0x00007fffb2a7c13b in ?? () from /usr/local/lib/python3.6/dist-packages/torch/lib/libgomp.so.1
#9  0x00007fffb2a77b3c in GOMP_parallel_start () from /usr/local/lib/python3.6/dist-packages/torch/lib/libgomp.so.1
#10 0x00007fffd0f0c3bb in THFloatTensor_fill () from /usr/local/lib/python3.6/dist-packages/torch/lib/libTH.so.1
#11 0x00007fffe5946f10 in THPFloatTensor_fill_ (self=<torch.FloatTensor at remote 0x7ffff6a8a7c8>, args=<optimized out>, kwargs=<optimized out>) at /pytorch/torch/csrc/generic/TensorMethods.cpp:993
#12 0x00000000004c4e7b in _PyCFunction_FastCallDict (kwargs=0x0, nargs=140737331636168, args=0x7ffff6b54e50, func_obj=<built-in method fill_ of torch.FloatTensor object at remote 0x7ffff6a8a7c8>)
    at ../Objects/methodobject.c:231
#13 _PyCFunction_FastCallKeywords (func=func@entry=<built-in method fill_ of torch.FloatTensor object at remote 0x7ffff6a8a7c8>, stack=stack@entry=0x7ffff6b54e50, nargs=nargs@entry=1, kwnames=kwnames@entry=0x0)
    at ../Objects/methodobject.c:294
#14 0x000000000054f924 in call_function (pp_stack=pp_stack@entry=0x7fffffffdcc8, oparg=<optimized out>, kwnames=kwnames@entry=0x0) at ../Python/ceval.c:4824
#15 0x000000000055400f in _PyEval_EvalFrameDefault (f=<optimized out>, throwflag=<optimized out>) at ../Python/ceval.c:3322
#16 0x000000000054ea28 in PyEval_EvalFrameEx (throwflag=0, f=
    Frame 0x7ffff6b54cc0, for file /usr/local/lib/python3.6/dist-packages/torch/tensor.py, line 334, in __rdiv__ (self=<torch.FloatTensor at remote 0x7ffff6a8a748>, other=1)) at ../Python/ceval.c:753
#17 _PyFunction_FastCall (co=co@entry=0x7fffa7c4b660, args=<optimized out>, args@entry=0x7fffffffde20, nargs=nargs@entry=2, 
    globals=globals@entry={'__name__': 'torch.tensor', '__doc__': None, '__package__': 'torch', '__loader__': <SourceFileLoader(name='torch.tensor', path='/usr/local/lib/python3.6/dist-packages/torch/tensor.py') at remote 0x7fffa7c3eb70>, '__spec__': <ModuleSpec(name='torch.tensor', loader=<...>, origin='/usr/local/lib/python3.6/dist-packages/torch/tensor.py', loader_state=None, submodule_search_locations=None, _set_fileattr=True, _cached='/usr/local/lib/python3.6/dist-packages/torch/__pycache__/tensor.cpython-36.pyc', _initializing=False) at remote 0x7fffa7c3eba8>, '__file__': '/usr/local/lib/python3.6/dist-packages/torch/tensor.py', '__cached__': '/usr/local/lib/python3.6/dist-packages/torch/__pycache__/tensor.cpython-36.pyc', '__builtins__': {'__name__': 'builtins', '__doc__': "Built-in functions, exceptions, and other objects.\n\nNoteworthy: None is the `nil' object; Ellipsis represents `...' in slices.", '__package__': '', '__loader__': <type at remote 0xaa0b98>, '__spec__': <ModuleSpec(name='builtins', loader=<type at remote 0xa...(truncated)) at ../Python/ceval.c:4906
#18 0x0000000000558832 in _PyFunction_FastCallDict (func=func@entry=<function at remote 0x7fffa7c4f158>, args=args@entry=0x7fffffffde20, nargs=nargs@entry=2, kwargs=kwargs@entry=0x0) at ../Python/ceval.c:5008
#19 0x0000000000459fb1 in _PyObject_FastCallDict (kwargs=0x0, nargs=2, args=0x7fffffffde20, func=<function at remote 0x7fffa7c4f158>) at ../Objects/abstract.c:2310
#20 _PyObject_Call_Prepend (func=<function at remote 0x7fffa7c4f158>, obj=<optimized out>, args=(1,), kwargs=0x0) at ../Objects/abstract.c:2373
#21 0x0000000000459a3e in PyObject_Call (func=func@entry=<method at remote 0x7ffff6c66b08>, args=args@entry=(1,), kwargs=kwargs@entry=0x0) at ../Objects/abstract.c:2261
#22 0x00000000004e1208 in call_maybe (o=<torch.FloatTensor at remote 0x7ffff6a8a748>, nameid=<optimized out>, format=0x7716aa "(O)", format=0x7716aa "(O)") at ../Objects/typeobject.c:1493
#23 0x0000000000455083 in binary_op1 (op_slot=240, w=<torch.FloatTensor at remote 0x7ffff6a8a748>, v=1) at ../Objects/abstract.c:786
#24 binary_op (op_name=0x66065a "/", op_slot=240, w=<torch.FloatTensor at remote 0x7ffff6a8a748>, v=1, v@entry=0x0) at ../Objects/abstract.c:809
#25 PyNumber_TrueDivide (v=v@entry=1, w=w@entry=<torch.FloatTensor at remote 0x7ffff6a8a748>) at ../Objects/abstract.c:992
#26 0x0000000000552993 in _PyEval_EvalFrameDefault (f=<optimized out>, throwflag=<optimized out>) at ../Python/ceval.c:1434
```

### PR DESCRIPTION
**Pull request by zou3519**
Fixes #4093

There was a bug where users could pass in indices to the sparse tensor constructor that imply tensors that have larger sizes than the size that was passed to the constructor. This adds error checking for that.

### CODE REVIEW

File: aten/src/THS/generic/THSTensor.c [hunk 1]
```diff
@@ -167,6 +167,26 @@ THSTensor *THSTensor_(newWithTensorAndSize)(THLongTensor *indices, THTensor *val
   else {
     THArgCheck(THLongStorage_size(sizes) == nDimI + nDimV, 2,
         "number of dimensions must be nDimI + nDimV");
+
+    THLongTensor *max_indices = THLongTensor_new();
```
Comment (soumith) [inline aten/src/THS/generic/THSTensor.c]: `max_indices` is never freed. This is a memory leak.

**Review by soumith — CHANGES_REQUESTED** [review]

**Comment by SsnL** [comment]
Looks good, but please fix the merge conflicts. 

We should really move all of sparse tensor into ATen one day, i.e.. having _indices and _values both be ATen tensors.

**Comment by zou3519** [comment]
Rebased. The onnx-fb-univese check seems to have failed because a build script won't run?

**Pull request merged**